added catching of php7 fatal exceptions in application #20110

bozerkins opened the pull request:
| Q | A |
| --- | --- |
| Branch? | ~~2.8~~ 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #20110 |
| License | MIT |
| Doc PR | \- - - |
- added new private method to application, which would process the exception and return the exit code / throw it
- added another catch into command run method, which would ensure that under PHP7 all exceptions are being catched as well

bozerkins committed:
added catching of php7 fatal exceptions

- added new private method to application, which would process the exception and return the exit code / throw it
- added another catch into command run method, which would ensure that under PHP7 all exceptions are being catched as well

nicolas-grekas commented:
Doesn't #19813 fix the issue?

bozerkins commented:
partially.

when no dispatcher is present - the fatal exceptions will not be catched anyways. 

https://github.com/fonsecas72/symfony/blob/65e53ece4c867dc0447d3a215fdda8170e163745/src/Symfony/Component/Console/Application.php#L845

Tobion commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -156,6 +140,39 @@ public function run(InputInterface $input = null, OutputInterface $output = null
+        if ($output instanceof ConsoleOutputInterface) {
+            $this->renderException($e, $output->getErrorOutput());
+        } else {
+            $this->renderException($e, $output);
```
The phpdoc of `renderException` with `@param \Exception      $e` is not correct anymore and needs to be updated too.

bozerkins replied:
done some changes. took the approach of #19813

now should be just fine

bozerkins committed:
added catching throwable exceptions when running console application

- implemented compatibilit with PHP7 fatal errors
- using FatalThrowableError for BC with PHP5

bozerkins committed:
minor code tweak

lyrixx commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -120,20 +120,28 @@ public function run(InputInterface $input = null, OutputInterface $output = null
+        $exitCode = null;
         try {
             $exitCode = $this->doRun($input, $output);
         } catch (\Exception $e) {
```
Could could rename  `$e` in `$exception` and remove the inner block.

nicolas-grekas commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -120,20 +120,28 @@ public function run(InputInterface $input = null, OutputInterface $output = null
+        if ($exception) {
             if (!$this->catchExceptions) {
-                throw $e;
+                throw $exception;
```
$e should be thrown here, not $exception (no wrapper)

bozerkins replied:
you mean FatalThrowableError should not be here at all?

nicolas-grekas replied:
yes, when rethrowing, the original exception should be thrown

nicolas-grekas commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -120,20 +120,28 @@ public function run(InputInterface $input = null, OutputInterface $output = null
         $this->configureIO($input, $output);
 
+        $exception = null;
+        $exitCode = null;
```
is this line required?

bozerkins replied:
the intend was to make code more readable and easy to understand.

but sure, this can be removed

nicolas-grekas commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -106,7 +106,7 @@ public function setDispatcher(EventDispatcherInterface $dispatcher)
      * @return int 0 if everything went fine, or an error code
      *
-     * @throws \Exception When doRun returns Exception
+     * @throws \Exception When doRun returns Exception or Throwable
```
the whole line should be removed imho, we don't document "rethrowing"

bozerkins committed:
rethrowing original exception extend of wrapper over Error exception

bozerkins commented:
done. now the changes have become as simple and straight forward as possible

lyrixx commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -123,6 +123,10 @@ public function run(InputInterface $input = null, OutputInterface $output = null
+        } catch (\Throwable $e) {
+        }
+
+        if ($e) {
```
This does not work. `if (isset($e))`.

bozerkins committed:
check if exception variable isset

nicolas-grekas commented:
Let's hange the type hint on the renderException method also (Throwable instead of Exception), that would be consistent with what is done here

stof commented:
@nicolas-grekas we cannot for BC reasons (it is common for projects to create their own Application class extending the Symfony one, and they may overwrite the method)

stof commented:
ah no, there is no typehint currently, so it will not break PHP

nicolas-grekas commented:
It's only a doc comment, not a php one, still not allowed?

nicolas-grekas commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -123,6 +123,10 @@ public function run(InputInterface $input = null, OutputInterface $output = null
         try {
             $exitCode = $this->doRun($input, $output);
         } catch (\Exception $e) {
+        } catch (\Throwable $e) {
```
Line 109 should still be removed

nicolas-grekas replied:
On 2.7, so in another PR, nevermind this comment :)

nicolas-grekas commented:
Here is the patch I propose t finish this PR:
```diff
--- a/src/Symfony/Component/Console/Application.php
+++ b/src/Symfony/Component/Console/Application.php
@@ -102,8 +102,6 @@ class Application
      * @param OutputInterface $output An Output instance
      *
      * @return int 0 if everything went fine, or an error code
-     *
-     * @throws \Exception When doRun returns Exception
      */
     public function run(InputInterface $input = null, OutputInterface $output = null)
     {
@@ -634,7 +632,7 @@ class Application
     /**
      * Renders a caught exception.
      *
-     * @param \Exception      $e      An exception instance
+     * @param \Throwable      $e      An exception instance
      * @param OutputInterface $output An OutputInterface instance
      */
     public function renderException($e, $output)
@@ -831,8 +829,6 @@ class Application
      * @param OutputInterface $output  An Output instance
      *
      * @return int 0 if everything went fine, or an error code
-     *
-     * @throws \Exception when the command being run threw an exception
      */
     protected function doRunCommand(Command $command, InputInterface $input, OutputInterface $output)
     {
```

Note also that it should be merged on 2.7!

fabpot commented:
@bozerkins Can you take @nicolas-grekas comments into account?
